design/metadata-handling.md

```diff
@@ -0,0 +1,417 @@
+- Generating the metadata or the networkdata should work regardless whether the
+  other template was provided.
+- Documentation of this feature would be very important.
+- The controllers must be able to recreate all status objects, i.e. no
+  necessary information should be stored in the status without being recoverable
+
+### Risks and Mitigations
+
+From a user point of view, if this is not well documented or understood, it
+could lead to unexpected errors or behaviors. The errors have to be easily
+understandable.
+
+## Design Details
+
+BareMetalHost would have an added field `metaData` in the `Spec` pointing to a
+secret containing the metadata map. Bare Metal Operator would then pass this map
+to Ironic.
+
+The Metal3Machine will be modified as follow:
+
+```yaml
+apiVersion: infrastructure.cluster.x-k8s.io/v1alpha4
+kind: Metal3Machine
+metadata:
+  name: machine-1
+  namespace: default
+spec:
+  dataTemplate:
+    name: nodepool-1
+    namespace: default
+```
+
+or alternatively
+
+```yaml
+apiVersion: infrastructure.cluster.x-k8s.io/v1alpha4
+kind: Metal3Machine
+metadata:
+  name: machine-1
+  namespace: default
+spec:
+  metaData:
+    name: machine-1-metadata
+    namespace: default
+  networkData:
+    name: machine-1-networkData
+    namespace: default
+```
+
+A `metaData` field would be added, consisting of a secret reference containing
+the name and the namespace of the secret containing the metaData for this
+Metal3Machine. A `networkData` field would be added, consisting of a secret
+reference containing the name and the namespace of the secret containing the
+network configuration for this Metal3Machine.
+
+A `dataTemplate` field would be added, consisting of an object reference
+containing the name and the namespace of the secret containing the templates for
+the metadata and network data generation for this Metal3Machine. If this is set
+but either the `metaData` or `networkData` field is unset, then the
+Metal3Machine controller will wait until those are populated, depending on which
+templates were provided.
+
+The secret containing the metaData or the network data could be provided by the
+user directly.
+
+When CAPM3 controller will set the different fields in the BareMetalHost,
+it will reference the metadata secret and the network data secret
+in the BareMetalHost. If `metaData` or `networkData` field is unset, it will
+also remain unset on the BareMetalHost.
+
+When the Metal3Machine gets deleted, the CAPM3 controller will remove its
+references from the data template object. This will trigger the deletion of the
+secrets generated for this machine.
+
+A new object would be created, a Metal3DataTemplate type.
+
+```yaml
+apiVersion: infrastructure.cluster.x-k8s.io/v1alpha4
+kind: Metal3DataTemplate
+metadata:
+  name: nodepool-1
+  namespace: default
+  ownerReferences:
+  - apiVersion: infrastructure.cluster.x-k8s.io/v1alpha4
+    controller: true
+    kind: Metal3Cluster
+    name: cluster-1
+spec:
+  metaData: |
```

Comment:
So the content of metaData and networkData are strings that should contain embedded YAML or JSON content? Is metaData always YAML and networkData always JSON, as in this example?

Reply:
yes, the format must be like this

Reply:
That's going to be very difficult to validate, and probably use.

Do you know any of the values that need to be templated in advance? Could we define a CRD with more of an API for setting those values, and have it generate the YAML or JSON to put into the secret and pass to the host?

For example, a structure like this would let the kubernetes API service do some of the validation for us.

```
metaData:
  - fieldName: abc
    literalValue: def
  - fieldName: local-hostname
    dataSource: "{{ machineName }}"
  - fieldName: index
    dataSource: "{{ index }}"
    dataType: int
networkData:
  links:
    - - fieldName: id
        literalValue: enp1s0
      - fieldName: type
        literalValue: phy
      - fieldName: ethernet_mac_address
        dataSource: "{{ bareMetalHostMACByName \"eth0\" }}"
    - - fieldName: id
        literalValue: enp2s0
      - fieldName: type
        literalValue: phy
      - fieldName: ethernet_mac_address
        dataSource: "{{ bareMetalHostMACByName \"eth1\" }}"
```

The network data section would be even better if it just used the fields directly. Something like

```
networkData:
  links:
    - id: enp1s0
      type: phy
      ethernetMacAddress: "{{ bareMetalHostMACByName \"eth0\" }}"
    - id: enp2s0
      type: phy
      ethernetMacAddress: "{{ bareMetalHostMACByName \"eth1\" }}"
```

The implementation of that form would be a little more complex, because everything might have to be treated as a possible source of template data. But maybe not all fields need to be templates? Or maybe it looks more like

```
networkData:
  links:
    - id: enp1s0
      type: phy
      ethernetMacAddressSource: eth0
    - id: enp2s0
      type: phy
      ethernetMacAddressSource: eth1
```

Reply:
This is a very good idea for the validation. We could also take an approach of letting the user define a map directly for the metadata, so it will be validated as a map[string]string. For the network, I also like the approach, though I would not bake in the functions, since the user might want to directly give the mac address for example. 

I am a bit unsure if we would take away some features from the network data. Considering the definition https://docs.openstack.org/nova/latest/_downloads/9119ca7ac90aa2990e762c08baea3a36/network_data.json there are different types of links, networks etc. Either we need to implement it so that we validate again all the fields, so re-implementing the nova specs, or we leave more room for the user, with the opportunity for mistakes